File: src/dataviewer/backend/src/api/models/detection.py
"""
Detection Pydantic models for YOLO11 object detection system.

These models define the request/response schemas for object detection
endpoints and match the frontend TypeScript type definitions.
"""

from pydantic import BaseModel, Field


class DetectionRequest(BaseModel):
    """Request parameters for running object detection."""

    frames: list[int] | None = Field(
        default=None,
        description="Specific frame indices to process. If None, processes all frames.",
    )
    confidence: float = Field(
        default=0.25,
        ge=0.0,
        le=1.0,
        description="Minimum confidence threshold for detections.",
    )
    model: str = Field(
        default="yolo11n",
        description="YOLO model variant: yolo11n, yolo11s, yolo11m, yolo11l, yolo11x",
    )


class Detection(BaseModel):
    """Single object detection result."""

    class_id: int = Field(ge=0, description="COCO class ID")
    class_name: str = Field(description="Human-readable class name")
    confidence: float = Field(ge=0.0, le=1.0, description="Detection confidence score")
    bbox: tuple[float, float, float, float] = Field(description="Bounding box as (x1, y1, x2, y2) in pixels")


class DetectionResult(BaseModel):
    """Detection results for a single frame."""

    frame: int = Field(ge=0, description="Frame index")
    detections: list[Detection] = Field(default_factory=list)
    processing_time_ms: float = Field(ge=0.0, description="Inference time in milliseconds")


class ClassSummary(BaseModel):
    """Summary statistics for a detection class."""

    count: int = Field(ge=0, description="Total detections of this class")
    avg_confidence: float = Field(ge=0.0, le=1.0, description="Average confidence")


class EpisodeDetectionSummary(BaseModel):
    """Complete detection results for an episode."""

    total_frames: int = Field(ge=0, description="Total frames in episode")
    processed_frames: int = Field(ge=0, description="Number of frames processed")
    total_detections: int = Field(ge=0, description="Total detections across all frames")
    detections_by_frame: list[DetectionResult] = Field(default_factory=list)
    class_summary: dict[str, ClassSummary] = Field(
        default_factory=dict,
        description="Detection statistics by class name",
    )

File: data-management/viewer/frontend/src/types/detection.ts
/**
 * TypeScript types for YOLO11 object detection.
 */

/**
 * Request parameters for running object detection.
 */
export interface DetectionRequest {
  /** Specific frame indices to process. If undefined, processes all frames. */
  frames?: number[]
  /** Minimum confidence threshold (0.0-1.0). Uses the backend-configured default when omitted. */
  confidence?: number
  /**
   * YOLO model variant. Closed-vocabulary defaults are `yolo11*`; open-vocabulary
   * variants such as `yolov8s-world` are selected automatically when `labels` is set.
   */
  model?: string
  /**
   * Optional open-vocabulary class names. When provided the backend switches to a
   * YOLO-World model and only returns boxes matching these labels.
   */
  labels?: string[]
  /**
   * Camera/stream key to source frames from (e.g. `observation.images.color`). When
   * omitted the backend uses the episode's first available camera.
   */
  camera?: string
}

/**
 * Single object detection result.
 */
export interface Detection {
  /** COCO class ID */
  classId: number
  /** Human-readable class name */
  className: string
  /** Detection confidence score (0.0-1.0) */
  confidence: number
  /** Bounding box as [x1, y1, x2, y2] in pixels */
  bbox: [number, number, number, number]
}

/**
 * Detection results for a single frame.
 */
export interface DetectionResult {
  /** Frame index */
  frame: number
  /** Detections found in this frame */
  detections: Detection[]
  /** Inference time in milliseconds */
  processingTimeMs: number
}

/**
 * Summary statistics for a detection class.
 */
export interface ClassSummary {
  /** Total detections of this class */
  count: number
  /** Average confidence */
  avgConfidence: number
}

/**
 * Complete detection results for an episode.
 */
export interface EpisodeDetectionSummary {
  /** Total frames in episode */
  totalFrames: number
  /** Number of frames processed */
  processedFrames: number
  /** Total detections across all frames */
  totalDetections: number
  /** Detection results by frame */
  detectionsByFrame: DetectionResult[]
  /** Detection statistics by class name */
  classSummary: Record<string, ClassSummary>
}
